src/transformers/models/gptsan_japanese/modeling_gptsan_japanese.py

```diff
@@ -0,0 +1,1410 @@
+# Otherwise, no mask at token positions from 0 to prefix_area_len, and regular causal mask thereafter.
+#
+# ex, total_seq=6, output_seq=6;
+# prefix_area_len=0  prefix_area_len=3  prefix_area_len=5
+# | 1 0 0 0 0 0 |    | 1 1 1 0 0 0 |    | 1 1 1 1 1 0 |
+# | 1 1 0 0 0 0 |    | 1 1 1 0 0 0 |    | 1 1 1 1 1 0 |
+# | 1 1 1 0 0 0 |    | 1 1 1 0 0 0 |    | 1 1 1 1 1 0 |
+# | 1 1 1 1 0 0 |    | 1 1 1 1 0 0 |    | 1 1 1 1 1 0 |
+# | 1 1 1 1 1 0 |    | 1 1 1 1 1 0 |    | 1 1 1 1 1 0 |
+# | 1 1 1 1 1 1 |    | 1 1 1 1 1 1 |    | 1 1 1 1 1 1 |
+#
+# Normally, the length of the prefix is set according to the position of the separator_token inserted by the tokenizer.
+# Since the prefix part refers back and forth, it can also work as an MLM. The part after that can be used for text generation as LM.
+#
+# ex;
+# input tokens are: x_token = tokenizer.encode("ぁぃぅぇ")
+# mask is;
+# SOT | 1 0 0 0 0 0 |
+# SEG | 1 1 0 0 0 0 |
+# ぁ  | 1 1 1 0 0 0 |
+# ぃ  | 1 1 1 1 0 0 |
+# ぅ  | 1 1 1 1 1 0 |
+# ぇ  | 1 1 1 1 1 1 |
+#
+# input tokens are: x_token = tokenizer.encode("", prefix_text="ぁぃぅぇ")
+# mask is;
+# SOT | 1 1 1 1 1 0 |
+# ぁ  | 1 1 1 1 1 0 |
+# ぃ  | 1 1 1 1 1 0 |
+# ぅ  | 1 1 1 1 1 0 |
+# ぇ  | 1 1 1 1 1 0 |
+# SEG | 1 1 1 1 1 1 |
+#
+# input tokens are: x_token = tokenizer.encode("ぅぇ", prefix_text="ぁぃ")
+# mask is;
+# SOT | 1 1 1 0 0 0 |
+# ぁ  | 1 1 1 0 0 0 |
+# ぃ  | 1 1 1 0 0 0 |
+# SEG | 1 1 1 1 0 0 |
+# ぅ  | 1 1 1 1 1 0 |
+# ぇ  | 1 1 1 1 1 1 |
+def make_variable_prefix_area_mask(total_seq, output_seq, prefix_area_len):
+    r"""
+    Args:
+        total_seq (`int`) :
+            inputs to token length.
+        output_seq (`int`) :
+            inputs to output token length.
+        prefix_area_len (`torch.Tensor` of shape [batch_size]) :
+            inputs to prefix area length of each batch.
+    Returns:
+        torch.Tensor[batch_size, 1, total_seq, output_seq]
+    """
+    device = prefix_area_len.device
+    inp_index = torch.arange(total_seq)[:, None].to(device)
+    out_index = torch.arange(output_seq).to(device)
+    m = inp_index >= out_index - output_seq + total_seq
+    lm_mask = m.float()  # language model mask
+    lm_mask = torch.reshape(lm_mask, [total_seq, output_seq])
+    # lm_mask shuld be [sequence, sequence]
+    weight = torch.transpose(torch.arange(output_seq)[:, None].to(device) < prefix_area_len, 1, 0)
+    weight = weight.float()  # mask for prefix area
+    # weight shuld be [batch, sequence]
+    mlm_mask = torch.reshape(weight, [1, -1, output_seq]).float()
+    mlm_ones = torch.ones(size=[total_seq, 1, 1], dtype=torch.float32).to(device)
+    mlm_mask = mlm_ones * mlm_mask
+    mlm_mask = torch.transpose(mlm_mask, 1, 0)
+    # mlm_mask shuld be [batch, sequence, sequence]
+    mask = ((lm_mask + mlm_mask) > 0).float()  # marge prefix area to causal mask
+    # mask shuld be [batch, sequence, sequence]
+    return mask.unsqueeze(dim=1)  # [batch, 1, sequence, sequence]
```

Comment:
I think this would better fit in the `tokenizer` file, where the mask would be created and just passed down accordingly. Let me just dig a bit deeper her! Since it might be the first time we have support for Prefix LM, we need to setup a good example

Reply:
I changed to use token_type_ids.
Locations with token_type_ids of 1 are treated as prefix input (that is, as input that can refer to the preceding and following tokens).
More token_type_ids in model input. Tokenizer also added token_type_ids to model_input_names.
I hope you like this.

Reply:
I LOVE it ! Great job! 👌